File: apps/mobile/android/androidApp/src/main/kotlin/cc/edecan/app/vm/AutomatizacionesViewModel.kt
package cc.edecan.app.vm

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import cc.edecan.shared.ApiException
import cc.edecan.shared.Automation
import cc.edecan.shared.AutomationRun
import cc.edecan.shared.EdecanApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class AutomatizacionesUiState(
    val cargando: Boolean = false,
    val automatizaciones: List<Automation> = emptyList(),
    val errorLista: String? = null,
    val creando: Boolean = false,
    val errorCrear: String? = null,
    /** ids con un `PATCH enabled` en vuelo — el Switch de esa fila queda
     * deshabilitado mientras tanto (`AutomatizacionesScreen`). */
    val idsEnCambio: Set<String> = emptySet(),
    val seleccionId: String? = null,
    val corridas: List<AutomationRun> = emptyList(),
    val cargandoCorridas: Boolean = false,
    val errorCorridas: String? = null,
)

/**
 * Estado y lógica de la pestaña "Automatizaciones" (`/v1/automations`,
 * `ROADMAP_V2.md` §7.4/§7.6/§7.10, WP-V5-07): lista con Switch optimista
 * (revierte a mano si el `PATCH` falla), alta simple (siempre agenda +
 * instrucción, ver `EdecanApi.createAutomation`) y detalle de corridas
 * (`GET /v1/automations/{id}/runs`).
 */
class AutomatizacionesViewModel : ViewModel() {
    private val _uiState = MutableStateFlow(AutomatizacionesUiState())
    val uiState: StateFlow<AutomatizacionesUiState> = _uiState.asStateFlow()

    private var yaCargado = false

    fun cargar(api: EdecanApi, forzar: Boolean = false) {
        if (yaCargado && !forzar) return
        yaCargado = true
        viewModelScope.launch {
            _uiState.update { it.copy(cargando = true, errorLista = null) }
            try {
                val automatizaciones = api.listAutomations()
                _uiState.update { it.copy(cargando = false, automatizaciones = automatizaciones) }
            } catch (e: ApiException) {
                _uiState.update { it.copy(cargando = false, errorLista = e.message) }
            }
        }
    }

    /** Alta simple: siempre `kind = "schedule"` (ver `EdecanApi.createAutomation`
     * y "AutomatizacionesScreen" para los presets de `rrule`). */
    fun crear(api: EdecanApi, nombre: String, rrule: String, instruccion: String) {
        val nombreLimpio = nombre.trim()
        val instruccionLimpia = instruccion.trim()
        if (nombreLimpio.isEmpty() || instruccionLimpia.isEmpty() || _uiState.value.creando) return
        viewModelScope.launch {
            _uiState.update { it.copy(creando = true, errorCrear = null) }
            try {
                val creada = api.createAutomation(nombre = nombreLimpio, rrule = rrule, instruccion = instruccionLimpia)
                _uiState.update {
                    it.copy(creando = false, automatizaciones = listOf(creada) + it.automatizaciones)
                }
            } catch (e: ApiException) {
                _uiState.update { it.copy(creando = false, errorCrear = e.message) }
            }
        }
    }

    /** Switch optimista: aplica `habilitado` en el estado local YA, dispara
     * el `PATCH` de verdad, y si falla revierte la fila a su valor anterior
     * (nunca deja el Switch "mintiendo" sobre lo que el servidor guardó). */
    fun alternar(api: EdecanApi, automation: Automation, habilitado: Boolean) {
        val id = automation.id
        if (id in _uiState.value.idsEnCambio) return
        _uiState.update { estado ->
            estado.copy(
                automatizaciones = estado.automatizaciones.map { a -> if (a.id == id) a.copy(enabled = habilitado) else a },
                idsEnCambio = estado.idsEnCambio + id,
                errorLista = null,
            )
        }
        viewModelScope.launch {
            try {
                val actualizada = api.toggleAutomation(id, habilitado)
                _uiState.update { estado ->
                    estado.copy(
                        automatizaciones = estado.automatizaciones.map { a -> if (a.id == id) actualizada else a },
                        idsEnCambio = estado.idsEnCambio - id,
                    )
                }
            } catch (e: ApiException) {
                _uiState.update { estado ->
                    estado.copy(
                        automatizaciones = estado.automatizaciones.map { a -> if (a.id == id) a.copy(enabled = !habilitado) else a },
                        idsEnCambio = estado.idsEnCambio - id,
                        errorLista = e.message,
                    )
                }
            }
        }
    }

    fun seleccionar(api: EdecanApi, automationId: String) {
        _uiState.update { it.copy(seleccionId = automationId, corridas = emptyList(), errorCorridas = null) }
        viewModelScope.launch {
            _uiState.update { it.copy(cargandoCorridas = true) }
            try {
                val corridas = api.listAutomationRuns(automationId)
                _uiState.update { it.copy(cargandoCorridas = false, corridas = corridas) }
            } catch (e: ApiException) {
                _uiState.update { it.copy(cargandoCorridas = false, errorCorridas = e.message) }
            }
        }
    }

    fun cerrarDetalle() {
        _uiState.update { it.copy(seleccionId = null, corridas = emptyList(), errorCorridas = null) }
    }
}

File: apps/mobile/ios/EdecanApp/Componentes/AutomatizacionesViewModel.swift
import Foundation
import Observation
import EdecanKit

/// Estado de ``AutomatizacionesView``/su detalle (`ARCHITECTURE.md` §11
/// `ROADMAP_V2.md` §7.4/§7.6/§7.10).
@MainActor
@Observable
final class AutomatizacionesViewModel {
    private(set) var automatizaciones: [AutomationOut] = []
    private(set) var cargando = false
    var errorLista: String?

    private(set) var runs: [AutomationRunOut] = []
    private(set) var cargandoRuns = false
    var errorRuns: String?

    private(set) var creando = false
    var errorCreacion: String?

    /// `id` de la automatización cuyo toggle está en vuelo — deshabilita
    /// SOLO esa fila mientras se confirma (o revierte) con el servidor.
    private(set) var idEnToggle: String?

    var sinDatosTodavia: Bool { automatizaciones.isEmpty && cargando }

    func cargar(client: APIClient?) async {
        guard let client else {
            errorLista = "No hay sesión activa."
            return
        }
        cargando = automatizaciones.isEmpty
        errorLista = nil
        defer { cargando = false }
        do {
            automatizaciones = try await client.listAutomations()
        } catch {
            errorLista = error.localizedDescription
        }
    }

    /// Optimista: refleja el nuevo `enabled` en la lista YA MISMO (mejor
    /// respuesta percibida al tocar el `Toggle`), y si el servidor lo
    /// rechaza (p. ej. `403` por el tope de automatizaciones activas del
    /// plan, `automations.py::_check_limit`) revierte la fila a su valor
    /// anterior y muestra el error.
    func alternar(id: String, enabled: Bool, client: APIClient?) async {
        guard let client, idEnToggle == nil else { return }
        guard let indice = automatizaciones.firstIndex(where: { $0.id == id }) else { return }
        let valorAnterior = automatizaciones[indice].enabled
        automatizaciones[indice].enabled = enabled
        idEnToggle = id
        errorLista = nil
        defer { idEnToggle = nil }
        do {
            let actualizada = try await client.toggleAutomation(id: id, enabled: enabled)
            if let i = automatizaciones.firstIndex(where: { $0.id == id }) {
                automatizaciones[i] = actualizada
            }
        } catch {
            if let i = automatizaciones.firstIndex(where: { $0.id == id }) {
                automatizaciones[i].enabled = valorAnterior
            }
            errorLista = error.localizedDescription
        }
    }

    /// "Alta simple": siempre un disparador de agenda (`rrule`) — ver
    /// `APIClient.createAutomation`.
    @discardableResult
    func crear(
        nombre: String, descripcion: String, rrule: String, instruccion: String, client: APIClient?
    ) async -> AutomationOut? {
        guard let client else {
            errorCreacion = "No hay sesión activa."
            return nil
        }
        let nombreLimpio = nombre.trimmingCharacters(in: .whitespacesAndNewlines)
        let instruccionLimpia = instruccion.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !nombreLimpio.isEmpty, !instruccionLimpia.isEmpty, !rrule.trimmingCharacters(in: .whitespaces).isEmpty,
              !creando
        else { return nil }
        creando = true
        errorCreacion = nil
        defer { creando = false }
        do {
            let creada = try await client.createAutomation(
                nombre: nombreLimpio, descripcion: descripcion, rrule: rrule, instruccion: instruccionLimpia
            )
            automatizaciones.insert(creada, at: 0)
            return creada
        } catch {
            errorCreacion = error.localizedDescription
            return nil
        }
    }

    func cargarRuns(id: String, client: APIClient?) async {
        guard let client else {
            errorRuns = "No hay sesión activa."
            return
        }
        cargandoRuns = true
        errorRuns = nil
        defer { cargandoRuns = false }
        do {
            runs = try await client.listAutomationRuns(id: id)
        } catch {
            errorRuns = error.localizedDescription
        }
    }
}
